### ISSUE TITLE
Log Injection breaks Error objects

### ISSUE THREAD

**Issue by mhassan1** [DataDog/dd-trace-js#931]
**Describe the bug**
When passing an `Error` object to `winston` (e.g. `log.error(new Error('oops'))`), the `log` plugin turns the `Error` into a plain `object` without the `message` and `stack` properties before passing it along to `winston`.

Here is a simple reproduction:
```
require('dd-trace').init({ logInjection: true })

const winston = require('winston')
const util = require('util')
const http = require('http')

const log = winston.createLogger({
  transports: [new winston.transports.Console()],
  format: winston.format.printf(util.inspect)
})

const app = require('express')()
app.get('/log', (req, res) => {
  log.error('good')
  log.error({ message: 'good' })
  log.error(new Error('not good'))
  res.send('')
})

app.listen(3003, () => {
  http.get('http://localhost:3003/log', () => {
    process.exit(0)
  })
})
```

Log output:
```
{
  dd: { trace_id: '1190699381420439375', span_id: '1245536781377984478' },
  message: 'good',
  level: 'error',
  [Symbol(level)]: 'error'
}
{
  dd: { trace_id: '1190699381420439375', span_id: '1245536781377984478' },
  message: 'good',
  level: 'error',
  [Symbol(level)]: 'error'
}
{
  dd: { trace_id: '1190699381420439375', span_id: '1245536781377984478' },
  level: 'error',
  [Symbol(level)]: 'error'
}
```

The third log entry is missing `message` because it is not iterable with `in`: https://github.com/DataDog/dd-trace-js/blob/d8d686736a5603c8925cef28b85879c3f37cd0ad/packages/dd-trace/src/plugins/util/log.js#L19

I feel like using `logInjection: true` should only augment log entries before passing them to `winston` and not modify them in a destructive way like this.

**Environment**

* **Operation system: OS X**
* **Node version: 12.14.0**
* **Tracer version: ?**
* **Agent version: ?**

**Comment by rochdev** [comment]
> I feel like using logInjection: true should only augment log entries before passing them to winston and not modify them in a destructive way like this.

This is what we were doing in the past, but it resulted in objects being altered in an unexpected way since the original object was mutated and caused all sorts of issues for users. Long story short, this is unfortunately something we cannot do.

I'll try to see if the instrumentation could be moved one layer down where we would have some kind of object created by Winston instead which would be different than the one passed by the user. Otherwise I'll need to find a way to copy this kind of object without losing information.

**Comment by mhassan1** [comment]
I could definitely see how mutating the object would cause issues for users, but I could also see how creating a new object that's similar but not the same might also cause issues for users.

A couple of ideas I had:
1. Let the user decide what they want the behavior to be:
```
logInjection: {
  // mutate log record when injecting `dd` attributes, instead of creating a shallow clone
  // this might be a good idea if you're passing Error objects
  mutateLogRecord: true
}
```
2. Use a library like [clone](https://www.npmjs.com/package/clone), which can clone lots of objects without loss of fidelity, although I know it is not very performant.

**Comment by rochdev** [comment]
What version of Winston are you using? From what I can tell by looking at Winston's code this should already be handled and the error would never be passed to the tracer.

**Comment by rochdev** [comment]
It also looks like Winston is already mutating the original object, so it may be ok to do it in the tracer since it means it cannot be avoided, at least with that logger.

**Comment by mhassan1** [comment]
I'm on `winston@3.1.0`. Here I see that it mutates the passed object only if it already has a `message` property; otherwise, it puts the passed object in `message`: https://github.com/winstonjs/winston/blob/73ae01f951600306242e00dd0d2b0a85b6d9d254/lib/winston/create-logger.js#L78

For example:
```
const e1 = { so: 'good' }
log.error(e1)
assert(!e1.level) // not mutated

const e2 = { message: 'good' }
log.error(e2)
assert(e2.level) // mutated

const e3 = new Error('not good')
log.error(e3)
assert(e3.level) // mutated
```

So `winston` is already mutating the original object if it has `.message` on it, and is already creating a new object if it doesn't. It seems like it would be safe for the tracer to always mutate the `record` for `winston@3`.

**Comment by mhassan1** [comment]
Also, `winston@2` puts the object in `meta` and does not mutate it, so the argument that `winston` is already mutating the object doesn't work for `winston@2`.

**Comment by rochdev** [comment]
It took longer than expected, but I was finally able to fix this completely for all loggers using different strategies:

* For `bunyan`, the log record is always a clone, so we can safely inject.
* For `pino`, the injection is now done right before serialization, and removed right after.
* For `winston` 1-2, the injection is done right before each transport is called, and removed right after.
* For `winston` >=3, the injection is done when a record is written to the logger stream, and removed right after.

This means that for all implementation, records are no longer cloned, mutations are avoided when possible, and objects are restored to their previous state when mutations are required. This should have the best performance while avoiding mutations.

Hopefully this should resolve this issue as well as any other potential issues related to cloning and mutations.

**Comment by mhassan1** [comment]
great idea! thanks for all the hard work.

**Comment by rochdev** [comment]
Fixed in 0.20.2
